### PR TITLE
SNOW-624893: Create performance test scripts

### PR DESCRIPTION
**Pull request by sfc-gh-yixie**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   SNOW-624893

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Use cProfile, memory pofile, and time.time to gather data.

### CODE REVIEW

File: tests/performance_test/client-performance-test.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    'password': '<password>',
+    'role': '<role>',
+    'database': '<database>',
+    'schema': '<schema>',
+    'warehouse': '<warehouse>',
+}
+```
+
+## Running Tests
+Change directory to `<Snowflake Python Project Dir>/tests/performance_test`
```
Comment (sfc-gh-jdu) [inline tests/performance_test/client-performance-test.md:31]: just to keep consistent with other directories, should we use `tests/perf`

File: tests/performance_test/client-performance-test.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+### Use script to take a glance at the running time and memory used
+1. Run `python perf_runner.py <api> <ncalls> -m -s`
+    - `api` is a function define in `perf_runner.py`.
```
Comment (sfc-gh-jdu) [inline tests/performance_test/client-performance-test.md:35]:
```suggestion
    - `api` is a function defined in `perf_runner.py`.
```

File: tests/performance_test/perf_runner.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+connection_parameters_path = str(Path(__file__).absolute().parent.parent)
+sys.path.append(connection_parameters_path)
+from parameters import CONNECTION_PARAMETERS  # noqa: E402
+
+logger = logging.getLogger("snowflake.snowpark")
+logger.setLevel(logging.ERROR)
+logger.addHandler(logging.StreamHandler())
+
+
+def generate_columns(n: int):
```
Comment (sfc-gh-jdu) [inline tests/performance_test/perf_runner.py:25]: let's also have return type hints?

File: tests/performance_test/perf_runner.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        api = args.api
+        func = eval(api)
+        if args.memory:
+            import memory_profiler
+
+            func = memory_profiler.profile(func)
+        t0 = time.time()
+        dataframe = func(session, args.ncalls)
+        t1 = time.time()
+        print("Client side time elapsed: ", t1 - t0)
```
Comment (sfc-gh-jdu) [inline tests/performance_test/perf_runner.py:112]: It would be great if we can get the average time elapsed?

**Review by sfc-gh-jdu — COMMENTED** [review]
This is awesome. Should we build a recurring tests runner on jenkins to run these performance tests, then we can inspect any performance regression.

File: tests/perf/perf_runner.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+from pathlib import Path
+from typing import Iterable, List
+
+from snowflake.snowpark import DataFrame
+from snowflake.snowpark.functions import col
+from snowflake.snowpark.session import Session
+
+connection_parameters_path = str(Path(__file__).absolute().parent.parent)
+sys.path.append(connection_parameters_path)
+from parameters import CONNECTION_PARAMETERS  # noqa: E402
```
[the remark this answers is not in the document]
Comment (sfc-gh-yixie) [reply tests/perf/perf_runner.py:18]: pytest does some magical. `python perf_runner.py` doesn't work without this.

File: tests/perf/perf_runner.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+
+def generate_columns(n: int) -> List[str]:
+    return [f'{i} as {"a" * 50}{i}' for i in range(n)]
```
[the remark this answers is not in the document]
Comment (sfc-gh-yixie) [reply tests/perf/perf_runner.py:22]: Yes it's to add length to column name. The characters shouldn't have much impact.

File: tests/perf/client-performance-test.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
+# Setting up and Running Performance Tests
```
Comment (sfc-gh-jdu) [inline tests/perf/client-performance-test.md:1]: Probably just name this file `README.md` so it will display automatically when going to `perf` directory

**Review by sfc-gh-jdu — APPROVED** [review]

File: tests/perf/client-performance-test.md (hunk 7 above)
Comment (sfc-gh-yixie) [reply tests/perf/client-performance-test.md]: Good suggestion. Updated.

**Pull request merged**